### PR TITLE
Expand documentation on CLI and workflows

### LINKED ISSUE #124: [FR] Expand documentation on CLI and workflows
**Is your feature request related to a problem? Please describe.**
Documentation can be expanded on workflows between the app and repo such as exporting existing rules and importing into the repo. In particular, using the CLI to facilitate these workflows could be annotated better.

### PR DESCRIPTION
**Pull request by brokensound77**
## Issues
resolves #124 

## Summary
Adds documentation for CLI usage and advanced workflows

### CODE REVIEW

File: CLI.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+This covers more advanced CLI use cases and workflows. To [get started](README.md#getting-started) with the CLI, reference 
+the [readme](README.md). Basic use of the CLI such as [creating a rule](CONTRIBUTING.md#creating-a-rule-with-the-cli) or 
+[testing](CONTRIBUTING.md#testing-a-rule-with-the-cli) are referenced in the [contribution guide](CONTRIBUTING.md).
+
+
+## Using a config file or environment variables
+
+CLI commands which are tied to Kibana and Elasticsearch are capable of parsing auth-related keyword args from a config 
+file or environment variables. Any command option which calls `callback=set_param_values` will search for set kwargs.
```
Comment (rw-access) [inline CLI.md:11]: I think we should assume the reader isn't familiar with the python here and spell things out as they apply to using the CLI

File: CLI.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+* config values
+* prompt (this only applies to certain values)
+
+#### Setup a config file
+
+In the root directory of this repo, create the file `.detection-rules-cfg.json` and add relevant values
+
+#### Using environment variables
+
+Environment variables using the key format: `DR_<UPPERCASED_KEY_NAME>` will be parsed in commands which expect it.
```
Comment (rw-access) [inline CLI.md:26]: change key to argument

File: CLI.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+## Importing rules into the repo
+
+You can import rules into the repo using the `create-rule` or `import-rules` commands. Both of these commands will 
+require that the rules are schema-compliant and able to pass full validation. The biggest benefit to using these 
+commands is that they will strip[*](#i18n-note) additional fields[**](#i18n-note-2) and prompt for missing required 
+fields. 
+
+Alternatively, you can manually place rule files in the directory and run tests to validate as well.
+
+<a id="i18n-note">\* Note</a>: This is currently limited to flat fields and may not apply to nested values.<br>
```
Comment (rw-access) [inline CLI.md:38]: what does this have to do with i18n?

File: CLI.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+<a id="i18n-note-2">\** Note</a>: Additional fields are based on the current schema at the time the command is used.
+
+
+#### `create-rule`
+
+This command will allow you to pass a rule file using the `-c/--config` parameter. This is limited to one rule at a time
```
Comment (rw-access) [inline CLI.md:44]: should we talk about standard usage first? maybe put the help text here?

File: CLI.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+In addition to the formats mentioned using `create-rule`, this will also accept an `.ndjson`/`jsonl` file 
+containing multiple rules (as would be the case with a bulk export).
+
+This will also strip additional fields and prompt for missing required fields.
+
+<a id="i18n-note">\* Note</a>: This will attempt to parse ALL files recursively within a specified directory.
+
+
+## Uploading rules to Kibana
```
Comment (rw-access) [inline CLI.md:65]: I partially regret adding this command for the testing/support surface area it introduced. I think we should add a note describing what the designed use case is: testing a rule before sharing back to the repo. It's just not complete and mature enough yet for me to encourage its use.

File: CLI.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+```
+
+_*Note that in order to load a custom rule, the proper index must be setup first. The simplest way to do this is to click 
+the `Load prebuilt detection rules and timeline templates` button on the `detections` page in the Kibana security app._
+
+
+## Converting between JSON and TOML
+
+[Importing rules](#importing-rules-into-the-repo) will convert from any supported format to toml. Additionally, the 
+command `view-rule` will also allow you to view a converted rule without importing it by specifying the `--as-rule` flag.
```
Comment (rw-access) [inline CLI.md:81]: I think this flag should be renamed (internally as well). View rule as rule feels like a tautology.

`--format repo/api` could be better

File: CLI.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+To view a rule in JSON format, you can also use the `view-rule` command with the `--as-api` flag, which is the default.
+(See the [note](#a-note-on-versioning-and-the-generated-json-formatted-rules) on the JSON formatted rules)
+
+Multiple rules can be converted into their JSON representation by using the `build-release` command. This will assemble
```
Comment (rw-access) [inline CLI.md:86]: I think we need a note here and in the CLI text that this command is made for our team in particular.

If we want to support JSON export, I'm not sure if this is the right way. I would rather just not create that functionality until someone asks for it.

File: CLI.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+To view a rule in JSON format, you can also use the `view-rule` command with the `--as-api` flag, which is the default.
+(See the [note](#a-note-on-versioning-and-the-generated-json-formatted-rules) on the JSON formatted rules)
+
+Multiple rules can be converted into their JSON representation by using the `build-release` command. This will assemble 
+rules which meet the specified filters in the [packages.yml](etc/packages.yml) file and stage them in the generated 
+`releases/{release}/rules`
+directory.
+
+## A note on versioning and the generated JSON formatted rules
```
Comment (rw-access) [inline CLI.md:91]: shorter section title. "Note on version handling" is fine

File: CLI.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+The rule toml files exist slightly different than they do in their final state as a json file in Kibana. The files are
```
Comment (rw-access) [inline CLI.md:93]: JSON

File: CLI.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+white space stripped, normalized, sorted, and indented, prior to their json conversion. Everything within the `metadata`
+table is also stripped out, as this is meant for rule development and management only.
```
Comment (rw-access) [inline CLI.md:95]: meant to be used only in the context of this repository and not in Kibana.

File: CLI.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+Additionally, the version of the rule is added to the file prior to exporting it. This is done to restrict version bumps
+to occur intentionally right before we create a release. Version bumps are auto-incremented based on detected changes in
```
Comment (rw-access) [inline CLI.md:98]: s/Version bumps/Versions

File: CLI.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+The rule toml files exist slightly different than they do in their final state as a json file in Kibana. The files are
+white space stripped, normalized, sorted, and indented, prior to their json conversion. Everything within the `metadata`
+table is also stripped out, as this is meant for rule development and management only.
+
+Additionally, the version of the rule is added to the file prior to exporting it. This is done to restrict version bumps
+to occur intentionally right before we create a release. Version bumps are auto-incremented based on detected changes in 
+rules. This is based on the hash of the rule in the following format:
+* sorted json
+* serialized
+* b64 encoded
```
Comment (rw-access) [inline CLI.md:102]: maybe too late now, but why the encoding?

File: README.md [hunk 13]
```diff
@@ -75,6 +75,7 @@ Commands:
 
 The [contribution guide](CONTRIBUTING.md) describes how to use the `create-rule` and `test` commands to create and test a new rule when contributing to Detection Rules.
 
+For more advanced CLI usage, refer to the [CLI guide](CLI.md).
```
Comment (rw-access) [inline README.md:78]: if CLI isn't already spelled out, we should do that here

File: CLI.md (hunk 3 above)
Comment (brokensound77) [reply CLI.md:38]: copy 🍝

File: CLI.md (hunk 5 above)
Comment (brokensound77) [reply CLI.md:65]: Agree but it is there and needs explanation. I like the idea of the note - will add

File: CLI.md (hunk 7 above)
Comment (brokensound77) [reply CLI.md:86]: sounds good

File: CLI.md (hunk 12 above)
Comment (brokensound77) [reply CLI.md:102]: easier to compare and troubleshoot

File: CLI.md (hunk 1 above)
Comment (brokensound77) [reply CLI.md]: standardized the args a bit better and added a list of them under `config` section of CLI doc

File: CLI.md [hunk 14]
```diff
@@ -0,0 +1,132 @@
+# Command Line Interface
+
+This covers more advanced CLI use cases and workflows. To [get started](README.md#getting-started) with the CLI, reference 
+the [readme](README.md). Basic use of the CLI such as [creating a rule](CONTRIBUTING.md#creating-a-rule-with-the-cli) or
```
Comment (rw-access) [inline CLI.md:4]:
```suggestion
the [README](README.md). Basic use of the CLI such as [creating a rule](CONTRIBUTING.md#creating-a-rule-with-the-cli) or 
```

File: CLI.md [hunk 15]
```diff
@@ -0,0 +1,132 @@
+# Command Line Interface
```
Comment (rw-access) [inline CLI.md:1]:
```suggestion
# Command Line Interface (CLI)
```

File: CLI.md [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+
+Toml formatted rule files can be uploaded as custom rules using the `kibana-upload` command. To upload more than one 
+file, specify multiple files at a time as individual args. This command is meant to support uploading and testing of 
+rules and is not intended for production use in its current state.
+
+```console
+python -m detection_rules kibana-upload my-rules/example_custom_rule.toml
+```
+
+_*Note that in order to load a custom rule, the proper index must be setup first. The simplest way to do this is to click
```
Comment (rw-access) [inline CLI.md:105]:
```suggestion
_*To load a custom rule, the proper index must be setup first. The simplest way to do this is to click 
```

File: detection_rules/eswrap.py [hunk 17]
```diff
@@ -177,7 +177,7 @@ def run(self, agent_hostname, indexes, verbose=True, **match):
 
 @es_group.command('collect-events')
 @click.argument('agent-hostname')
-@click.option('--host', callback=set_param_values, expose_value=True)
+@click.option('--es-url', callback=set_param_values, expose_value=True)
```
Comment (rw-access) [inline detection_rules/eswrap.py:180]: a `-u` could be good here

Comment (rw-access) [reply detection_rules/eswrap.py:180]: I was also wondering if spelling it out as `elasticsearch_url` would be better

File: detection_rules/main.py [hunk 18]
```diff
@@ -123,9 +123,9 @@ def mass_update(ctx, query, field):
 @root.command('view-rule')
 @click.argument('rule-id', required=False)
 @click.option('--rule-file', '-f', type=click.Path(dir_okay=False), help='Optionally view a rule from a specified file')
-@click.option('--as-api/--as-rule', default=True, help='Print the rule in final api or rule format')
+@click.option('--api-format/--rule-format', default=True, help='Print the rule in final api or rule format')
```
Comment (rw-access) [inline detection_rules/main.py:126]: thanks!
